Use stream_id in composed messages and recipient comparisons.

showell opened the pull request.

smarx commented:
*Automated message from Dropbox CLA bot*

@showell, it looks like you've already signed the Dropbox CLA. Thanks!

showell commented:
Hmm...I need a bit more testing on compose fade stuff.

showell commented:
Ok, this is ready to review now.

timabbott commented:
This has gotten to be an impressively long branch!    I merged through c6069bc9cd050167b731bc9a7b327fb6f47f3837.  

The device-related stuff, I think there's a pretty good chance we'll be showing in the future, but is easy to add back; and the last bit of stuff has some linter errors.

The main thing I'm worried about in the last bit is 4ca8388aa66d11988364cc40459b94ffa97fdb95, which seems to add O(n) work to single-user updates; have you tried creating 2K users with presence data and seeing if the "update 1 user's stuff" perf regresses there?  I think it's possible the new version is better, but am not sure, and since it's important to large sites, it seems worth checking.

timabbott commented:
Another thing to think about while you're working on this is the new drafts.js logic; it's possible that we can now convert that use to the user IDs (in the case they exist; I guess we need the fallback to emails).

showell commented:
The single-update stuff actually used to be O(NlogN), because we were re-sorting the entire list of users to figure out indexes for the new row.  Now we just linearly search for the right spot.  I tested with 2k users and the new code is faster.  We were and still are removing the row first, which is an O(N) lookup.

I fixed the lint errors and rebased.  I haven't looked at the drafts stuff yet, though, and I'm about to go to bed, and then I have jury duty tomorrow.

timabbott commented:
Ahh, OK, makes sense.  So as long as we're not doing something expensive as the operations, we should be good.

timabbott commented:
OK, that makes sense; merged!  

Re: drafts, it's just a more important piece of the issue about saving user ID rather than lists of emails for compose.  We can just do that when we get to it in the general purge of emails from the frontend.

When you get back, I think it may be high time for a round of frontend traceback squashing.
